File: srt_editor.rb
# srt_editor.rb
class Subtitle
  attr_accessor :index, :start, :end, :text
  def initialize(index, start, end_, text)
    @index = index
    @start = start  # секунды с миллисекундами (Float)
    @end = end_
    @text = text
  end
end

def parse_time(s)
  s = s.strip.tr(',', '.')
  h, m, sec = s.split(':')
  sec, ms = sec.split('.')
  h.to_i * 3600 + m.to_i * 60 + sec.to_i + (ms || 0).to_f / 1000
end

def format_time(t)
  h = t.to_i / 3600
  m = (t.to_i % 3600) / 60
  s = t.to_i % 60
  ms = ((t - t.to_i) * 1000).round
  format("%02d:%02d:%02d,%03d", h, m, s, ms)
end

def parse_srt(content)
  subs = []
  blocks = content.strip.split(/\n\s*\n/)
  blocks.each do |block|
    lines = block.lines.map(&:chomp)
    next if lines.size < 3
    idx = lines[0].to_i
    time_line = lines[1]
    from, to = time_line.split('-->').map(&:strip)
    start = parse_time(from)
    end_ = parse_time(to)
    text = lines[2..-1].join("\n")
    subs << Subtitle.new(idx, start, end_, text)
  end
  subs.each_with_index { |s, i| s.index = i+1 }
  subs
end

def format_srt(subs)
  subs.map do |s|
    "#{s.index}\n#{format_time(s.start)} --> #{format_time(s.end)}\n#{s.text}"
  end.join("\n\n") + "\n"
end

class Editor
  attr_reader :filename, :subs

  def initialize(filename)
    @filename = filename
    @subs = parse_srt(File.read(filename, encoding: 'utf-8'))
    puts "Загружено #{@subs.size} субтитров."
  end

  def save(fname = nil)
    fname ||= @filename
    File.write(fname, format_srt(@subs), encoding: 'utf-8')
    puts "Сохранено."
  end

  def shift(ms)
    delta = ms / 1000.0
    @subs.each do |s|
      s.start += delta
      s.end += delta
    end
    puts "Сдвиг на #{ms} мс выполнен."
  end

  def renumber
    @subs.each_with_index { |s, i| s.index = i+1 }
    puts "Перенумерация выполнена."
  end

  def validate
    errs = []
    if @subs.empty?
      errs << "Нет субтитров."
    else
      prev_end = 0
      @subs.each_with_index do |s, i|
        if s.start >= s.end
          errs << "##{i+1}: начало >= конец"
        end
        if i > 0 && s.start < prev_end
          errs << "##{i+1}: перекрытие с предыдущим"
        end
        prev_end = s.end
      end
    end
    errs
  end

  def list
    @subs.each do |s|
      puts "#{s.index}  #{format_time(s.start)} --> #{format_time(s.end)}\n#{s.text}\n"
    end
  end

  def add(start_str, end_str, text)
    start = parse_time(start_str)
    end_ = parse_time(end_str)
    @subs << Subtitle.new(@subs.size+1, start, end_, text)
    renumber
    puts "Субтитр добавлен."
  end

  def edit(idx, start_str, end_str, text)
    if idx < 1 || idx > @subs.size
      puts "Неверный номер."
      return
    end
    s = @subs[idx-1]
    s.start = parse_time(start_str) if start_str && !start_str.empty?
    s.end = parse_time(end_str) if end_str && !end_str.empty?
    s.text = text if text && !text.empty?
    puts "Субтитр обновлён."
  end

  def delete(idx)
    if idx < 1 || idx > @subs.size
      puts "Неверный номер."
      return
    end
    @subs.delete_at(idx-1)
    renumber
    puts "Субтитр удалён."
  end
end

if __FILE__ == $0
  if ARGV.empty?
    puts "Использование: ruby srt_editor.rb [--shift мс] [--check] [--output file] файл.srt"
    exit
  end
  filename = nil
  shift_ms = 0
  check_only = false
  output_file = nil
  i = 0
  while i < ARGV.size
    case ARGV[i]
    when '--shift'
      shift_ms = ARGV[i+1].to_i
      i += 2
    when '--check'
      check_only = true
      i += 1
    when '--output'
      output_file = ARGV[i+1]
      i += 2
    else
      filename = ARGV[i]
      i += 1
    end
  end
  unless filename
    puts "Не указан файл."
    exit
  end
  begin
    editor = Editor.new(filename)
  rescue => e
    puts "Ошибка загрузки: #{e.message}"
    exit
  end
  if check_only
    errs = editor.validate
    if errs.empty?
      puts "Субтитры корректны."
    else
      puts "Ошибки:"
      errs.each { |e| puts "  #{e}" }
    end
    exit
  end
  if shift_ms != 0
    editor.shift(shift_ms)
    editor.save(output_file || filename)
    exit
  end

  puts "Редактор субтитров. Введите help для списка команд."
  loop do
    print "> "
    cmd = gets.chomp.strip
    next if cmd.empty?
    parts = cmd.split
    op = parts[0]
    case op
    when 'quit'
      break
    when 'help'
      puts "list, show <номер>, add, edit <номер> [время_начала] [время_конца] [текст], delete <номер>, shift <мс>, renumber, validate, save, quit"
    when 'list'
      editor.list
    when 'show'
      if parts.size < 2
        puts "Укажите номер."
        next
      end
      idx = parts[1].to_i
      if idx < 1 || idx > editor.subs.size
        puts "Неверный номер."
        next
      end
      s = editor.subs[idx-1]
      puts "#{s.index}  #{format_time(s.start)} --> #{format_time(s.end)}\n#{s.text}"
    when 'add'
      print "Начало (HH:MM:SS,mmm): "
      start = gets.chomp
      print "Конец (HH:MM:SS,mmm): "
      end_ = gets.chomp
      puts "Текст (введите END для завершения):"
      lines = []
      while (line = gets.chomp) != 'END'
        lines << line
      end
      editor.add(start, end_, lines.join("\n"))
    when 'edit'
      if parts.size < 2
        puts "Укажите номер."
        next
      end
      idx = parts[1].to_i
      print "Новое начало (оставьте пустым): "
      start = gets.chomp
      print "Новый конец (оставьте пустым): "
      end_ = gets.chomp
      print "Новый текст (оставьте пустым): "
      text = gets.chomp
      editor.edit(idx, start, end_, text)
    when 'delete'
      if parts.size < 2
        puts "Укажите номер."
        next
      end
      idx = parts[1].to_i
      editor.delete(idx)
    when 'shift'
      if parts.size < 2
        puts "Укажите сдвиг в мс."
        next
      end
      ms = parts[1].to_i
      editor.shift(ms)
    when 'renumber'
      editor.renumber
    when 'validate'
      errs = editor.validate
      if errs.empty?
        puts "OK"
      else
        puts "Ошибки:"
        errs.each { |e| puts "  #{e}" }
      end
    when 'save'
      fname = parts[1] if parts.size > 1
      editor.save(fname)
    else
      puts "Неизвестная команда."
    end
  end
end

File: srt_editor.js
// srt_editor.js
const fs = require('fs');
const readline = require('readline');

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

function parseTime(s) {
    s = s.trim().replace(',', '.');
    const parts = s.split(':');
    if (parts.length !== 3) return 0;
    const h = parseInt(parts[0]);
    const m = parseInt(parts[1]);
    const secParts = parts[2].split('.');
    const sec = parseInt(secParts[0]);
    const ms = secParts.length > 1 ? parseInt(secParts[1]) : 0;
    return h * 3600 + m * 60 + sec + ms / 1000;
}

function formatTime(t) {
    const h = Math.floor(t / 3600);
    const m = Math.floor((t % 3600) / 60);
    const s = Math.floor(t % 60);
    const ms = Math.round((t - Math.floor(t)) * 1000);
    return `${String(h).padStart(2,'0')}:${String(m).padStart(2,'0')}:${String(s).padStart(2,'0')},${String(ms).padStart(3,'0')}`;
}

function parseSRT(content) {
    const subs = [];
    const blocks = content.trim().split(/\n\s*\n/);
    for (const block of blocks) {
        const lines = block.split('\n');
        if (lines.length < 3) continue;
        let idx = parseInt(lines[0]);
        const timeLine = lines[1];
        const parts = timeLine.split('-->');
        if (parts.length !== 2) continue;
        const start = parseTime(parts[0]);
        const end = parseTime(parts[1]);
        const text = lines.slice(2).join('\n');
        subs.push({ index: idx, start, end, text });
    }
    subs.forEach((s, i) => s.index = i+1);
    return subs;
}

function formatSRT(subs) {
    return subs.map(s => `${s.index}\n${formatTime(s.start)} --> ${formatTime(s.end)}\n${s.text}`).join('\n\n') + '\n';
}

class Editor {
    constructor(filename) {
        this.filename = filename;
        this.subs = parseSRT(fs.readFileSync(filename, 'utf8'));
        console.log(`Загружено ${this.subs.length} субтитров.`);
    }
    save(fname) {
        fname = fname || this.filename;
        fs.writeFileSync(fname, formatSRT(this.subs), 'utf8');
        console.log('Сохранено.');
    }
    shift(ms) {
        const delta = ms / 1000;
        this.subs.forEach(s => { s.start += delta; s.end += delta; });
        console.log(`Сдвиг на ${ms} мс выполнен.`);
    }
    renumber() {
        this.subs.forEach((s, i) => s.index = i+1);
        console.log('Перенумерация выполнена.');
    }
    validate() {
        const errs = [];
        if (this.subs.length === 0) errs.push('Нет субтитров.');
        let prevEnd = 0;
        this.subs.forEach((s, i) => {
            if (s.start >= s.end) errs.push(`#${i+1}: начало >= конец`);
            if (i > 0 && s.start < prevEnd) errs.push(`#${i+1}: перекрытие с предыдущим`);
            prevEnd = s.end;
        });
        return errs;
    }
    list() {
        this.subs.forEach(s => {
            console.log(`#${s.index}  ${formatTime(s.start)} --> ${formatTime(s.end)}\n${s.text}\n`);
        });
    }
    add(startStr, endStr, text) {
        const start = parseTime(startStr);
        const end = parseTime(endStr);
        this.subs.push({ index: this.subs.length+1, start, end, text });
        this.renumber();
        console.log('Субтитр добавлен.');
    }
    edit(idx, startStr, endStr, text) {
        if (idx < 1 || idx > this.subs.length) { console.log('Неверный номер.'); return; }
        const s = this.subs[idx-1];
        if (startStr && startStr.trim()) s.start = parseTime(startStr);
        if (endStr && endStr.trim()) s.end = parseTime(endStr);
        if (text && text.trim()) s.text = text;
        console.log('Субтитр обновлён.');
    }
    delete(idx) {
        if (idx < 1 || idx > this.subs.length) { console.log('Неверный номер.'); return; }
        this.subs.splice(idx-1, 1);
        this.renumber();
        console.log('Субтитр удалён.');
    }
}

function main() {
    const args = process.argv.slice(2);
    if (args.length === 0) {
        console.log('Использование: node srt_editor.js [--shift мс] [--check] [--output file] файл.srt');
        return;
    }
    let filename = null;
    let shiftMs = 0;
    let checkOnly = false;
    let outputFile = null;
    for (let i=0; i<args.length; i++) {
        if (args[i] === '--shift' && i+1 < args.length) {
            shiftMs = parseInt(args[++i]);
        } else if (args[i] === '--check') {
            checkOnly = true;
        } else if (args[i] === '--output' && i+1 < args.length) {
            outputFile = args[++i];
        } else {
            filename = args[i];
        }
    }
    if (!filename) {
        console.log('Не указан файл.');
        return;
    }
    let editor;
    try {
        editor = new Editor(filename);
    } catch (e) {
        console.log('Ошибка загрузки:', e.message);
        return;
    }
    if (checkOnly) {
        const errs = editor.validate();
        if (errs.length === 0) console.log('Субтитры корректны.');
        else {
            console.log('Ошибки:');
            errs.forEach(e => console.log('  ' + e));
        }
        return;
    }
    if (shiftMs !== 0) {
        editor.shift(shiftMs);
        editor.save(outputFile || filename);
        return;
    }

    console.log('Редактор субтитров. Введите help для списка команд.');
    const prompt = () => {
        rl.question('> ', (cmd) => {
            cmd = cmd.trim();
            if (!cmd) { prompt(); return; }
            const parts = cmd.split(/\s+/);
            const op = parts[0];
            switch (op) {
                case 'quit': rl.close(); return;
                case 'help':
                    console.log('list, show <номер>, add, edit <номер> [время_начала] [время_конца] [текст], delete <номер>, shift <мс>, renumber, validate, save, quit');
                    break;
                case 'list': editor.list(); break;
                case 'show':
                    if (parts.length < 2) { console.log('Укажите номер.'); break; }
                    const idx = parseInt(parts[1]);
                    if (idx<1 || idx>editor.subs.length) { console.log('Неверный номер.'); break; }
                    const s = editor.subs[idx-1];
                    console.log(`#${s.index}  ${formatTime(s.start)} --> ${formatTime(s.end)}\n${s.text}`);
                    break;
                case 'add': {
                    rl.question('Начало (HH:MM:SS,mmm): ', (start) => {
                        rl.question('Конец (HH:MM:SS,mmm): ', (end) => {
                            console.log('Текст (введите END для завершения):');
                            let lines = [];
                            const readText = () => {
                                rl.question('', (line) => {
                                    if (line === 'END') {
                                        editor.add(start, end, lines.join('\n'));
                                        prompt();
                                    } else {
                                        lines.push(line);
                                        readText();
                                    }
                                });
                            };
                            readText();
                        });
                    });
                    return;
                }
                case 'edit':
                    if (parts.length < 2) { console.log('Укажите номер.'); break; }
                    const idx2 = parseInt(parts[1]);
                    rl.question('Новое начало (оставьте пустым): ', (start) => {
                        rl.question('Новый конец (оставьте пустым): ', (end) => {
                            rl.question('Новый текст (оставьте пустым): ', (text) => {
                                editor.edit(idx2, start, end, text);
                                prompt();
                            });
                        });
                    });
                    return;
                case 'delete':
                    if (parts.length < 2) { console.log('Укажите номер.'); break; }
                    editor.delete(parseInt(parts[1]));
                    break;
                case 'shift':
                    if (parts.length < 2) { console.log('Укажите сдвиг в мс.'); break; }
                    editor.shift(parseInt(parts[1]));
                    break;
                case 'renumber': editor.renumber(); break;
                case 'validate':
                    const errs = editor.validate();
                    if (errs.length === 0) console.log('OK');
                    else {
                        console.log('Ошибки:');
                        errs.forEach(e => console.log('  ' + e));
                    }
                    break;
                case 'save':
                    const fname = parts.length > 1 ? parts[1] : null;
                    editor.save(fname);
                    break;
                default: console.log('Неизвестная команда.');
            }
            prompt();
        });
    };
    prompt();
}

main();
